pkg/etcdcli/etcdcli.go

```diff
@@ -0,0 +1,180 @@
+		nodeListerSynced:      kubeInformers.InformersFor("").Core().V1().Nodes().Informer().HasSynced,
+		endpointsListerSynced: kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Informer().HasSynced,
+	}
+}
+
+func (g *etcdClientGetter) getEtcdClient() (*clientv3.Client, error) {
+	if !g.nodeListerSynced() {
+		return nil, fmt.Errorf("node lister not synced")
+	}
+	if !g.endpointsListerSynced() {
+		return nil, fmt.Errorf("node lister not synced")
+	}
+
+	etcdEndpoints := []string{}
+	nodes, err := g.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
+	for _, node := range nodes {
+		internalIP, err := getInternalIPAddressForNodeName(node)
+		if err != nil {
+			return nil, err
+		}
+		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", internalIP))
+	}
+
+	hostEtcd, err := g.endpointsLister.Endpoints(operatorclient.TargetNamespace).Get("host-etcd")
+	if err != nil {
+		return nil, err
+	}
+	for _, addr := range hostEtcd.Subsets[0].Addresses {
+		if addr.Hostname == "etcd-bootstrap" {
+			// etcd-bootstrap has a valid IP in host-etcd
+			etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", addr.IP))
+			break
+		}
+	}
+
+	c, err := getEtcdClient(etcdEndpoints)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func getEtcdClient(endpoints []string) (*clientv3.Client, error) {
+	dialOptions := []grpc.DialOption{
+		grpc.WithBlock(), // block until the underlying connection is up
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      "/var/run/secrets/etcd-client/tls.crt",
+		KeyFile:       "/var/run/secrets/etcd-client/tls.key",
+		TrustedCAFile: "/var/run/configmaps/etcd-ca/ca-bundle.crt",
+	}
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	cfg := &clientv3.Config{
+		DialOptions: dialOptions,
+		Endpoints:   endpoints,
+		DialTimeout: 15 * time.Second,
+		TLS:         tlsConfig,
+	}
+
+	cli, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cli, err
+}
+
+func getInternalIPAddressForNodeName(node *corev1.Node) (string, error) {
+	for _, currAddress := range node.Status.Addresses {
+		if currAddress.Type == corev1.NodeInternalIP {
+			return currAddress.Address, nil
+		}
+	}
+	return "", fmt.Errorf("node/%s missing %s", node.Name, corev1.NodeInternalIP)
+}
+
+func (g *etcdClientGetter) MemberAdd(peerURL string) (bool, error) {
+	cli, err := g.getEtcdClient()
+	if err != nil {
+		return false, err
+	}
+	defer cli.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	membersResp, err := cli.MemberList(ctx)
+	cancel()
+	if err != nil {
+		return false, err
+	}
+
+	for _, member := range membersResp.Members {
+		if len(member.PeerURLs) > 0 && member.PeerURLs[0] == peerURL {
+			return true, err
+		}
+	}
+
+	_, err = cli.MemberAdd(ctx, []string{peerURL})
+	if err != nil {
+		return false, err
+	}
+	return true, err
+}
+
+func (g *etcdClientGetter) MemberRemove(member string) (bool, error) {
```

Comment:
go doc the bool

Reply:
whether or not the operation was successful or not, it can be assumed with `err != nil`, removing it